# Seasonal Agriculture Performance Analysis

**Project Type:** Major Project — VOIS For Tech / AICTE Data Analytics Internship  
**Batch:** August 2026–2027  
**Dataset:** seasonal_agriculture_performance_dataset.csv

---

## 1. Introduction

Agricultural activities are influenced by seasonal variations in environmental conditions, farming practices, resource availability, and market conditions. Understanding how agricultural performance changes across seasons is essential for effective planning and resource allocation.

This project analyzes a dataset of agricultural activities carried out across different seasons, geographical areas, and farming conditions. The dataset contains information related to farming practices, environmental conditions, crop production, resource usage, and economic performance.

The goal is to investigate seasonal differences in agricultural performance by identifying meaningful patterns, trends, relationships, and variations within the available data.

## 2. Problem Statement

Raw agricultural data does not clearly explain how agricultural performance changes across seasons or what patterns can be observed in different seasonal conditions.

**The problem is to:**
- Analyze the given agricultural dataset
- Investigate seasonal differences in agricultural performance
- Identify meaningful patterns, trends, relationships, and variations
- Provide evidence-based findings and recommendations

## 3. Objectives

1. Explore and understand the agricultural dataset
2. Clean and prepare the data for analysis
3. Examine how agricultural performance varies across seasons
4. Identify important seasonal patterns and trends
5. Investigate relationships between environmental conditions and agricultural outcomes
6. Compare performance across irrigation methods and crops
7. Apply appropriate statistical and visualization techniques
8. Develop meaningful conclusions and data-driven recommendations

## 4. Dataset Description

The dataset contains **4,000 records** of agricultural farm data with **28 columns** covering:

| Category | Columns |
|---|---|
| **Geography** | Farm_ID, State, District |
| **Crop & Season** | Crop, Season |
| **Farm Size** | Farm_Area_Hectares |
| **Environmental** | Rainfall_mm, Avg_Temperature_C, Humidity_pct, Sunlight_Hours_Day |
| **Soil** | Soil_pH, Soil_Moisture_pct |
| **Nutrients** | Nitrogen_kg_ha, Phosphorus_kg_ha, Potassium_kg_ha |
| **Farming Practices** | Irrigation_Method, Fertilizer_kg_ha, Pesticide_Litre_ha, Seed_Quality_Score |
| **Performance** | Yield_Tonnes_Ha, Production_Tonnes |
| **Economics** | Market_Price_INR_Tonne, Total_Cost_INR, Revenue_INR, Profit_INR |
| **Water** | Water_Used_m3, Water_Efficiency_t_per_1000m3 |
| **Risk** | Disease_Pest_Risk_pct |

## 5. Import Libraries

In [1]:
import pandas as pd
import numpy as np
import matplotlib
matplotlib.use('Agg')
import matplotlib.pyplot as plt
import seaborn as sns
import warnings
import os

warnings.filterwarnings('ignore')

# Set plotting style
sns.set_style('whitegrid')
plt.rcParams['figure.figsize'] = (10, 6)
plt.rcParams['font.size'] = 12
plt.rcParams['axes.titlesize'] = 14
plt.rcParams['axes.labelsize'] = 12

# Create charts directory
os.makedirs('charts', exist_ok=True)

print("Libraries imported successfully.")

Libraries imported successfully.


## 6. Load Dataset

In [2]:
df = pd.read_csv('seasonal_agriculture_performance_dataset.csv')
print(f"Dataset loaded successfully.")
print(f"Shape: {df.shape[0]} rows × {df.shape[1]} columns")

Dataset loaded successfully.
Shape: 4000 rows × 28 columns


## 7. Data Understanding

Let's explore the dataset structure, data types, missing values, and basic statistics.

### 7.1 Dataset Shape and Column Names

In [3]:
print(f"Number of rows: {df.shape[0]}")
print(f"Number of columns: {df.shape[1]}")
print(f"\nColumn names:")
for i, col in enumerate(df.columns, 1):
    print(f"  {i}. {col}")

Number of rows: 4000
Number of columns: 28

Column names:
  1. Farm_ID
  2. State
  3. District
  4. Crop
  5. Season
  6. Farm_Area_Hectares
  7. Rainfall_mm
  8. Avg_Temperature_C
  9. Humidity_pct
  10. Sunlight_Hours_Day
  11. Soil_pH
  12. Soil_Moisture_pct
  13. Nitrogen_kg_ha
  14. Phosphorus_kg_ha
  15. Potassium_kg_ha
  16. Irrigation_Method
  17. Fertilizer_kg_ha
  18. Pesticide_Litre_ha
  19. Seed_Quality_Score
  20. Yield_Tonnes_Ha
  21. Production_Tonnes
  22. Market_Price_INR_Tonne
  23. Total_Cost_INR
  24. Revenue_INR
  25. Profit_INR
  26. Water_Used_m3
  27. Water_Efficiency_t_per_1000m3
  28. Disease_Pest_Risk_pct


### 7.2 Data Types

In [4]:
print(df.dtypes)

Farm_ID                              str
State                                str
District                             str
Crop                                 str
Season                               str
Farm_Area_Hectares               float64
Rainfall_mm                      float64
Avg_Temperature_C                float64
Humidity_pct                     float64
Sunlight_Hours_Day               float64
Soil_pH                          float64
Soil_Moisture_pct                float64
Nitrogen_kg_ha                   float64
Phosphorus_kg_ha                 float64
Potassium_kg_ha                  float64
Irrigation_Method                    str
Fertilizer_kg_ha                 float64
Pesticide_Litre_ha               float64
Seed_Quality_Score               float64
Yield_Tonnes_Ha                  float64
Production_Tonnes                float64
Market_Price_INR_Tonne             int64
Total_Cost_INR                     int64
Revenue_INR                        int64
Profit_INR      

### 7.3 First 5 Rows

In [5]:
df.head()

,Farm_ID,State,District,Crop,Season,Farm_Area_Hectares,Rainfall_mm,Avg_Temperature_C,Humidity_pct,Sunlight_Hours_Day,...,Seed_Quality_Score,Yield_Tonnes_Ha,Production_Tonnes,Market_Price_INR_Tonne,Total_Cost_INR,Revenue_INR,Profit_INR,Water_Used_m3,Water_Efficiency_t_per_1000m3,Disease_Pest_Risk_pct
0,SF10001,Andhra Pradesh,Rajkot,Wheat,Kharif,0.53,486.4,24.3,69.1,5.8,...,0.76,2.46,1.30,23700,42662,30810,-11852,237,5.485,55.3
1,SF10002,Maharashtra,Nalgonda,Maize,Kharif,6.53,855.4,25.7,78.4,5.1,...,0.94,0.30,1.96,20613,492351,40401,-451950,4953,0.396,49.9
2,SF10003,Telangana,Warangal,Pulses,Rabi,4.86,455.6,23.9,56.4,10.3,...,0.98,0.52,2.53,75283,315210,190466,-124744,1275,1.984,33.7
3,SF10004,Telangana,Indore,Rice,Kharif,4.50,753.2,29.7,65.2,6.3,...,0.74,1.84,8.28,24244,296444,200740,-95704,3834,2.160,50.6
4,SF10005,Karnataka,Ludhiana,Maize,Zaid,4.21,101.6,34.1,55.2,6.5,...,0.90,2.21,9.30,20818,337959,193607,-144352,3287,2.829,32.3


### 7.4 Missing Values

In [6]:
missing = df.isnull().sum()
missing_pct = (df.isnull().sum() / len(df) * 100).round(2)
missing_df = pd.DataFrame({'Missing Count': missing, 'Missing %': missing_pct})
missing_df = missing_df[missing_df['Missing Count'] > 0]
print("Columns with missing values:")
print(missing_df)
print(f"\nTotal missing values: {df.isnull().sum().sum()}")

Columns with missing values:
                   Missing Count  Missing %
Rainfall_mm                   48        1.2
Soil_Moisture_pct             40        1.0
Yield_Tonnes_Ha               32        0.8

Total missing values: 120


### 7.5 Duplicate Rows

In [7]:
duplicates = df.duplicated().sum()
print(f"Number of duplicate rows: {duplicates}")

Number of duplicate rows: 0


### 7.6 Unique Values of Categorical Columns

In [8]:
categorical_cols = ['State', 'District', 'Crop', 'Season', 'Irrigation_Method']
for col in categorical_cols:
    print(f"\n{col} ({df[col].nunique()} unique values):")
    print(f"  {sorted(df[col].unique())}")


State (8 unique values):
  ['Andhra Pradesh', 'Gujarat', 'Karnataka', 'Madhya Pradesh', 'Maharashtra', 'Punjab', 'Tamil Nadu', 'Telangana']

District (10 unique values):
  ['Erode', 'Guntur', 'Indore', 'Krishna', 'Ludhiana', 'Nalgonda', 'Nashik', 'Raichur', 'Rajkot', 'Warangal']

Crop (8 unique values):
  ['Chilli', 'Cotton', 'Groundnut', 'Maize', 'Pulses', 'Rice', 'Sugarcane', 'Wheat']

Season (3 unique values):
  ['Kharif', 'Rabi', 'Zaid']

Irrigation_Method (4 unique values):
  ['Drip', 'Flood', 'Rainfed', 'Sprinkler']


### 7.7 Descriptive Statistics

In [9]:
df.describe().round(2)

,Farm_Area_Hectares,Rainfall_mm,Avg_Temperature_C,Humidity_pct,Sunlight_Hours_Day,Soil_pH,Soil_Moisture_pct,Nitrogen_kg_ha,Phosphorus_kg_ha,Potassium_kg_ha,...,Seed_Quality_Score,Yield_Tonnes_Ha,Production_Tonnes,Market_Price_INR_Tonne,Total_Cost_INR,Revenue_INR,Profit_INR,Water_Used_m3,Water_Efficiency_t_per_1000m3,Disease_Pest_Risk_pct
count,4000.00,3952.00,4000.00,4000.00,4000.00,4000.00,3960.00,4000.00,4000.00,4000.00,...,4000.00,3968.00,4000.00,4000.00,4000.00,4000.00,4000.00,4000.00,4000.00,4000.00
mean,7.89,601.15,26.82,63.21,7.32,6.70,26.51,120.25,57.78,104.76,...,0.83,5.27,43.25,44714.17,526303.58,637860.05,111556.46,6045.57,5.39,46.37
std,4.22,288.93,3.82,11.96,1.12,0.64,6.71,31.22,17.02,27.74,...,0.10,13.36,126.88,30266.16,294419.14,636974.71,545081.89,5572.94,8.84,12.42
min,0.50,80.00,16.20,25.00,3.50,5.20,8.00,40.00,15.00,30.00,...,0.65,0.30,0.15,2594.00,26826.00,6107.00,-1019223.00,128.00,0.14,5.00
25%,4.26,370.20,23.90,54.78,6.60,6.26,21.70,98.50,46.20,86.20,...,0.74,1.05,5.02,21429.25,274577.50,207517.25,-148731.75,2268.75,1.66,37.30
50%,7.90,582.00,26.90,63.00,7.30,6.69,26.40,120.10,57.50,105.10,...,0.83,1.74,11.77,25913.00,519082.50,456830.50,3673.00,4435.00,2.94,46.40
75%,11.65,834.95,29.60,71.90,8.10,7.14,31.30,141.22,69.40,123.60,...,0.91,2.84,24.36,68279.00,761996.50,836874.75,216187.75,7829.75,5.15,54.90
max,15.00,1395.20,39.70,95.00,11.00,8.40,47.00,220.00,120.00,200.00,...,1.00,101.44,1424.40,131240.00,1349990.00,5080900.00,4345421.00,40902.00,79.80,88.90


### 7.8 Numerical Column Ranges

In [10]:
num_cols = df.select_dtypes(include='number').columns
range_df = pd.DataFrame({
    'Min': df[num_cols].min(),
    'Max': df[num_cols].max(),
    'Mean': df[num_cols].mean().round(2),
    'Median': df[num_cols].median().round(2)
})
print(range_df.to_string())

                                       Min         Max       Mean     Median
Farm_Area_Hectares                   0.500       15.00       7.89       7.90
Rainfall_mm                         80.000     1395.20     601.15     582.00
Avg_Temperature_C                   16.200       39.70      26.82      26.90
Humidity_pct                        25.000       95.00      63.21      63.00
Sunlight_Hours_Day                   3.500       11.00       7.32       7.30
Soil_pH                              5.200        8.40       6.70       6.69
Soil_Moisture_pct                    8.000       47.00      26.51      26.40
Nitrogen_kg_ha                      40.000      220.00     120.25     120.10
Phosphorus_kg_ha                    15.000      120.00      57.78      57.50
Potassium_kg_ha                     30.000      200.00     104.76     105.10
Fertilizer_kg_ha                    40.000      400.00     186.04     186.25
Pesticide_Litre_ha                   0.500       12.08       5.07       5.12

## 8. Data Cleaning

Based on the data understanding phase, we identified:
- **Rainfall_mm**: 48 missing values (1.2%)
- **Soil_Moisture_pct**: 40 missing values (1.0%)
- **Yield_Tonnes_Ha**: 32 missing values (0.8%)
- **No duplicate rows**
- **No invalid data types**

**Cleaning decisions:**
1. Fill missing numerical values with the **median** of their respective columns (robust to outliers)
2. No rows need to be dropped since missing values are few and can be imputed
3. The original CSV is not modified — all cleaning is done on the working DataFrame

In [11]:
# Store original counts for verification
original_shape = df.shape
missing_before = df.isnull().sum().sum()

# Fill missing values with median
columns_with_missing = ['Rainfall_mm', 'Soil_Moisture_pct', 'Yield_Tonnes_Ha']
for col in columns_with_missing:
    median_val = df[col].median()
    count_missing = df[col].isnull().sum()
    df[col] = df[col].fillna(median_val)
    print(f"Filled {count_missing} missing values in '{col}' with median = {median_val}")

missing_after = df.isnull().sum().sum()
print(f"\nMissing values before cleaning: {missing_before}")
print(f"Missing values after cleaning: {missing_after}")
print(f"Dataset shape unchanged: {df.shape == original_shape}")

Filled 48 missing values in 'Rainfall_mm' with median = 582.0


Filled 40 missing values in 'Soil_Moisture_pct' with median = 26.4
Filled 32 missing values in 'Yield_Tonnes_Ha' with median = 1.74

Missing values before cleaning: 120
Missing values after cleaning: 0
Dataset shape unchanged: True


### 8.1 Verify Data After Cleaning

In [12]:
# Confirm no missing values remain
print("Missing values after cleaning:")
print(df.isnull().sum()[df.isnull().sum() > 0])
if df.isnull().sum().sum() == 0:
    print("\n✓ No missing values remain. Dataset is clean.")

Missing values after cleaning:
Series([], dtype: int64)

✓ No missing values remain. Dataset is clean.


## 9. Exploratory Data Analysis

We will investigate the following analytical questions:

1. How does agricultural yield vary across seasons?
2. How does production vary across seasons?
3. How does profitability vary across seasons?
4. How do rainfall and temperature differ across seasons?
5. How does water usage vary across seasons?
6. How does water efficiency vary across seasons?
7. How does disease/pest risk vary across seasons?
8. How do different crops perform?
9. How do irrigation methods relate to yield and water efficiency?
10. What relationships exist between environmental conditions and agricultural outcomes?

## 10. Season-wise Analysis

### 10.1 Season Distribution

In [13]:
season_counts = df['Season'].value_counts()
print("Number of records per season:")
print(season_counts)
print(f"\nTotal: {season_counts.sum()}")

Number of records per season:
Season
Kharif    1779
Rabi      1627
Zaid       594
Name: count, dtype: int64

Total: 4000


### 10.2 Average Yield by Season

In [14]:
season_yield = df.groupby('Season')['Yield_Tonnes_Ha'].agg(['mean', 'median', 'std']).round(2)
season_yield.columns = ['Mean Yield (t/ha)', 'Median Yield (t/ha)', 'Std Dev']
print(season_yield)

# Chart 1: Average Yield by Season
fig, ax = plt.subplots(figsize=(8, 5))
colors = ['#2E86AB', '#A23B72', '#F18F01']
bars = ax.bar(season_yield.index, season_yield['Mean Yield (t/ha)'], color=colors, edgecolor='black', linewidth=0.5)
ax.set_title('Average Yield by Season', fontweight='bold', pad=15)
ax.set_xlabel('Season')
ax.set_ylabel('Average Yield (Tonnes/Hectare)')
for bar, val in zip(bars, season_yield['Mean Yield (t/ha)']):
    ax.text(bar.get_x() + bar.get_width()/2, bar.get_height() + 0.1, f'{val}', ha='center', fontsize=11)
plt.tight_layout()
plt.savefig('charts/01_avg_yield_by_season.png', dpi=150, bbox_inches='tight')
plt.show()
print("\nChart saved: charts/01_avg_yield_by_season.png")

        Mean Yield (t/ha)  Median Yield (t/ha)  Std Dev
Season                                                 
Kharif               5.63                 1.94    14.39
Rabi                 5.04                 1.67    12.75
Zaid                 4.64                 1.46    11.27



Chart saved: charts/01_avg_yield_by_season.png


**Observation:** Kharif season shows the highest average yield, followed by Rabi and Zaid.

**Interpretation:** This suggests that Kharif season conditions (higher rainfall, monsoon period) are associated with higher average yields in this dataset. However, the differences should be considered alongside the variation (standard deviation) within each season.

### 10.3 Average Profit by Season

In [15]:
season_profit = df.groupby('Season')['Profit_INR'].agg(['mean', 'median']).round(2)
season_profit.columns = ['Mean Profit (₹)', 'Median Profit (₹)']
print(season_profit)

# Profitable vs loss-making farms
print("\nProfitable vs Loss-making farms:")
for season in ['Kharif', 'Rabi', 'Zaid']:
    sdf = df[df['Season'] == season]
    profitable = (sdf['Profit_INR'] > 0).sum()
    loss = (sdf['Profit_INR'] <= 0).sum()
    pct = round(profitable / len(sdf) * 100, 1)
    print(f"  {season}: {profitable} profitable ({pct}%), {loss} loss-making")

# Chart 2: Average Profit by Season
fig, ax = plt.subplots(figsize=(8, 5))
profit_vals = season_profit['Mean Profit (₹)']
bar_colors = ['#2E86AB' if v >= 0 else '#E63946' for v in profit_vals]
bars = ax.bar(profit_vals.index, profit_vals.values, color=bar_colors, edgecolor='black', linewidth=0.5)
ax.set_title('Average Profit by Season', fontweight='bold', pad=15)
ax.set_xlabel('Season')
ax.set_ylabel('Average Profit (₹)')
ax.axhline(y=0, color='black', linewidth=0.8, linestyle='--')
for bar, val in zip(bars, profit_vals.values):
    ypos = bar.get_height() + 3000 if val >= 0 else bar.get_height() - 12000
    ax.text(bar.get_x() + bar.get_width()/2, ypos, f'₹{val:,.0f}', ha='center', fontsize=10)
plt.tight_layout()
plt.savefig('charts/02_avg_profit_by_season.png', dpi=150, bbox_inches='tight')
plt.show()
print("\nChart saved: charts/02_avg_profit_by_season.png")

        Mean Profit (₹)  Median Profit (₹)
Season                                    
Kharif        178914.65            38808.0
Rabi           87689.47            -3187.0
Zaid          -24804.82           -62144.0

Profitable vs Loss-making farms:
  Kharif: 1028 profitable (57.8%), 751 loss-making
  Rabi: 795 profitable (48.9%), 832 loss-making
  Zaid: 211 profitable (35.5%), 383 loss-making



Chart saved: charts/02_avg_profit_by_season.png


**Observation:** Kharif has the highest average profit (₹1,78,915), Rabi is moderate (₹87,689), and Zaid shows a negative average profit (-₹24,805).

**Interpretation:** Kharif season appears to be the most profitable on average. Zaid season's negative average profit indicates that more farms in this season incur losses than make profits in this dataset. This may be associated with the lower rainfall and different crop mix during Zaid.

### 10.4 Total Production by Season

In [16]:
season_production = df.groupby('Season')['Production_Tonnes'].agg(['sum', 'mean']).round(2)
season_production.columns = ['Total Production (Tonnes)', 'Mean Production (Tonnes)']
print(season_production)

# Chart 3: Total Production by Season
fig, ax = plt.subplots(figsize=(8, 5))
bars = ax.bar(season_production.index, season_production['Total Production (Tonnes)'], 
              color=['#2E86AB', '#A23B72', '#F18F01'], edgecolor='black', linewidth=0.5)
ax.set_title('Total Production by Season', fontweight='bold', pad=15)
ax.set_xlabel('Season')
ax.set_ylabel('Total Production (Tonnes)')
for bar, val in zip(bars, season_production['Total Production (Tonnes)']):
    ax.text(bar.get_x() + bar.get_width()/2, bar.get_height() + 500, f'{val:,.0f}', ha='center', fontsize=10)
plt.tight_layout()
plt.savefig('charts/03_total_production_by_season.png', dpi=150, bbox_inches='tight')
plt.show()
print("\nChart saved: charts/03_total_production_by_season.png")

        Total Production (Tonnes)  Mean Production (Tonnes)
Season                                                     
Kharif                   82387.61                     46.31
Rabi                     67498.88                     41.49
Zaid                     23098.35                     38.89



Chart saved: charts/03_total_production_by_season.png


**Observation:** Kharif has the highest total production, which is expected given it also has the most records (1,779). Even the mean production per farm is highest for Kharif.

**Note:** Total production comparisons should consider that seasons have unequal record counts.

## 11. Crop-wise Analysis

### 11.1 Crop-wise Average Yield

In [17]:
crop_yield = df.groupby('Crop')['Yield_Tonnes_Ha'].agg(['mean', 'median', 'count']).round(2)
crop_yield.columns = ['Mean Yield (t/ha)', 'Median Yield (t/ha)', 'Count']
crop_yield = crop_yield.sort_values('Mean Yield (t/ha)', ascending=False)
print(crop_yield)

# Chart 9: Crop-wise Average Yield
fig, ax = plt.subplots(figsize=(10, 6))
crop_colors = ['#264653', '#2A9D8F', '#E9C46A', '#F4A261', '#E76F51', '#606C38', '#283618', '#DDA15E']
bars = ax.barh(crop_yield.index[::-1], crop_yield['Mean Yield (t/ha)'][::-1], 
               color=crop_colors[:len(crop_yield)], edgecolor='black', linewidth=0.5)
ax.set_title('Crop-wise Average Yield', fontweight='bold', pad=15)
ax.set_xlabel('Average Yield (Tonnes/Hectare)')
for bar, val in zip(bars, crop_yield['Mean Yield (t/ha)'][::-1]):
    ax.text(bar.get_width() + 0.5, bar.get_y() + bar.get_height()/2, f'{val}', va='center', fontsize=10)
plt.tight_layout()
plt.savefig('charts/09_crop_wise_avg_yield.png', dpi=150, bbox_inches='tight')
plt.show()
print("\nChart saved: charts/09_crop_wise_avg_yield.png")

           Mean Yield (t/ha)  Median Yield (t/ha)  Count
Crop                                                    
Sugarcane              46.64                47.88    305
Maize                   2.71                 2.84    551
Rice                    2.43                 2.44    690
Wheat                   2.11                 2.22    614
Chilli                  1.54                 1.56    412
Groundnut               1.32                 1.40    424
Cotton                  1.23                 1.25    508
Pulses                  0.92                 0.94    496



Chart saved: charts/09_crop_wise_avg_yield.png


**Observation:** Sugarcane has a dramatically higher average yield (46.94 t/ha) compared to all other crops. This is expected since sugarcane is a high-tonnage crop by nature. Among other crops, Maize (2.72) and Rice (2.44) have relatively higher yields.

**Note:** Direct yield comparisons across crops should consider that different crops have inherently different yield ranges.

### 11.2 Crop-wise Average Profit

In [18]:
crop_profit = df.groupby('Crop')['Profit_INR'].agg(['mean', 'median']).round(2)
crop_profit.columns = ['Mean Profit (₹)', 'Median Profit (₹)']
crop_profit = crop_profit.sort_values('Mean Profit (₹)', ascending=False)
print(crop_profit)

           Mean Profit (₹)  Median Profit (₹)
Crop                                         
Sugarcane        817187.99           691166.0
Chilli           750878.34           536375.0
Cotton           124546.92            62092.5
Groundnut         44858.12            36307.0
Pulses            -4238.05             2575.5
Maize            -83978.33           -42625.0
Rice            -102213.50           -58286.0
Wheat           -123398.34           -76269.0


**Observation:** Sugarcane (₹8,17,188) and Chilli (₹7,50,878) show the highest average profits. Wheat (-₹1,23,398) and Rice (-₹1,02,214) show negative average profits in this dataset.

**Interpretation:** High-value crops (Sugarcane, Chilli) tend to generate more profit despite potentially higher costs. The negative average profit for Wheat and Rice suggests these crops face challenges in cost recovery within this dataset.

## 12. Environmental Analysis

### 12.1 Rainfall and Temperature by Season

In [19]:
env_season = df.groupby('Season').agg({
    'Rainfall_mm': 'mean',
    'Avg_Temperature_C': 'mean',
    'Humidity_pct': 'mean',
    'Sunlight_Hours_Day': 'mean'
}).round(2)
print("Environmental conditions by season:")
print(env_season)

Environmental conditions by season:
        Rainfall_mm  Avg_Temperature_C  Humidity_pct  Sunlight_Hours_Day
Season                                                                  
Kharif       849.20              28.45         71.81                6.79
Rabi         437.62              23.49         57.89                7.59
Zaid         304.65              31.04         52.01                8.18


### 12.2 Rainfall vs Yield

In [20]:
# Chart 4: Rainfall vs Yield (scatter)
fig, ax = plt.subplots(figsize=(10, 6))
season_colors = {'Kharif': '#2E86AB', 'Rabi': '#A23B72', 'Zaid': '#F18F01'}
for season, color in season_colors.items():
    mask = df['Season'] == season
    ax.scatter(df.loc[mask, 'Rainfall_mm'], df.loc[mask, 'Yield_Tonnes_Ha'], 
               alpha=0.3, s=15, color=color, label=season)
ax.set_title('Rainfall vs Yield', fontweight='bold', pad=15)
ax.set_xlabel('Rainfall (mm)')
ax.set_ylabel('Yield (Tonnes/Hectare)')
ax.legend(title='Season')
plt.tight_layout()
plt.savefig('charts/04_rainfall_vs_yield.png', dpi=150, bbox_inches='tight')
plt.show()

corr_rain_yield = df['Rainfall_mm'].corr(df['Yield_Tonnes_Ha'])
print(f"\nCorrelation between Rainfall and Yield: {corr_rain_yield:.3f}")
print("Chart saved: charts/04_rainfall_vs_yield.png")


Correlation between Rainfall and Yield: 0.031
Chart saved: charts/04_rainfall_vs_yield.png


**Observation:** The scatter plot shows no clear linear relationship between rainfall and yield. The correlation coefficient is very weak (close to 0).

**Interpretation:** Rainfall alone does not appear to be a strong predictor of yield in this dataset. Other factors (crop type, irrigation, farming practices) likely play more significant roles.

### 12.3 Temperature vs Yield

In [21]:
# Chart 5: Temperature vs Yield
fig, ax = plt.subplots(figsize=(10, 6))
for season, color in season_colors.items():
    mask = df['Season'] == season
    ax.scatter(df.loc[mask, 'Avg_Temperature_C'], df.loc[mask, 'Yield_Tonnes_Ha'], 
               alpha=0.3, s=15, color=color, label=season)
ax.set_title('Temperature vs Yield', fontweight='bold', pad=15)
ax.set_xlabel('Average Temperature (°C)')
ax.set_ylabel('Yield (Tonnes/Hectare)')
ax.legend(title='Season')
plt.tight_layout()
plt.savefig('charts/05_temperature_vs_yield.png', dpi=150, bbox_inches='tight')
plt.show()

corr_temp_yield = df['Avg_Temperature_C'].corr(df['Yield_Tonnes_Ha'])
print(f"\nCorrelation between Temperature and Yield: {corr_temp_yield:.3f}")
print("Chart saved: charts/05_temperature_vs_yield.png")


Correlation between Temperature and Yield: 0.010
Chart saved: charts/05_temperature_vs_yield.png


**Observation:** Temperature also shows a very weak correlation with yield. The scatter plot does not reveal a clear pattern.

**Interpretation:** Temperature alone is not a strong predictor of yield in this dataset. The seasonal clustering visible in the plot (Rabi = cooler, Zaid = hotter) reflects expected seasonal temperature differences.

## 13. Irrigation Analysis

### 13.1 Irrigation Method vs Yield

In [22]:
irr_yield = df.groupby('Irrigation_Method')['Yield_Tonnes_Ha'].agg(['mean', 'median', 'count']).round(2)
irr_yield.columns = ['Mean Yield', 'Median Yield', 'Count']
print(irr_yield)

# Chart 10: Irrigation Method vs Yield
fig, ax = plt.subplots(figsize=(8, 5))
irr_colors = ['#264653', '#2A9D8F', '#E9C46A', '#E76F51']
bars = ax.bar(irr_yield.index, irr_yield['Mean Yield'], color=irr_colors, edgecolor='black', linewidth=0.5)
ax.set_title('Average Yield by Irrigation Method', fontweight='bold', pad=15)
ax.set_xlabel('Irrigation Method')
ax.set_ylabel('Average Yield (Tonnes/Hectare)')
for bar, val in zip(bars, irr_yield['Mean Yield']):
    ax.text(bar.get_x() + bar.get_width()/2, bar.get_height() + 0.1, f'{val}', ha='center', fontsize=11)
plt.tight_layout()
plt.savefig('charts/10_irrigation_vs_yield.png', dpi=150, bbox_inches='tight')
plt.show()
print("\nChart saved: charts/10_irrigation_vs_yield.png")

                   Mean Yield  Median Yield  Count
Irrigation_Method                                 
Drip                     6.58          1.94    915
Flood                    4.86          1.66   1310
Rainfed                  4.60          1.72   1041
Sprinkler                5.16          1.78    734



Chart saved: charts/10_irrigation_vs_yield.png


### 13.2 Irrigation Method vs Water Efficiency

In [23]:
irr_water_eff = df.groupby('Irrigation_Method')['Water_Efficiency_t_per_1000m3'].agg(['mean', 'median']).round(2)
irr_water_eff.columns = ['Mean Water Efficiency', 'Median Water Efficiency']
print(irr_water_eff)

# Chart 11: Irrigation Method vs Water Efficiency
fig, ax = plt.subplots(figsize=(8, 5))
bars = ax.bar(irr_water_eff.index, irr_water_eff['Mean Water Efficiency'], 
              color=irr_colors, edgecolor='black', linewidth=0.5)
ax.set_title('Water Efficiency by Irrigation Method', fontweight='bold', pad=15)
ax.set_xlabel('Irrigation Method')
ax.set_ylabel('Water Efficiency (t per 1000 m³)')
for bar, val in zip(bars, irr_water_eff['Mean Water Efficiency']):
    ax.text(bar.get_x() + bar.get_width()/2, bar.get_height() + 0.1, f'{val}', ha='center', fontsize=11)
plt.tight_layout()
plt.savefig('charts/11_irrigation_vs_water_efficiency.png', dpi=150, bbox_inches='tight')
plt.show()
print("\nChart saved: charts/11_irrigation_vs_water_efficiency.png")

                   Mean Water Efficiency  Median Water Efficiency
Irrigation_Method                                                
Drip                                6.27                     3.26
Flood                               3.44                     1.95
Rainfed                             7.56                     4.58
Sprinkler                           4.67                     2.65



Chart saved: charts/11_irrigation_vs_water_efficiency.png


**Observation:** Drip irrigation shows the highest average yield (6.62 t/ha), while Rainfed shows the highest water efficiency (7.56 t/1000m³).

**Interpretation:** Drip irrigation is associated with better yields, possibly due to controlled water delivery. Rainfed farming shows higher water efficiency because it uses less measured water (relying on rainfall instead), so the efficiency metric naturally appears higher. This distinction is important — high water efficiency in rainfed does not necessarily mean better water management, it reflects lower measured water usage.

## 14. Economic Analysis

### 14.1 Revenue, Cost, and Profit by Season

In [24]:
econ_season = df.groupby('Season').agg({
    'Revenue_INR': 'mean',
    'Total_Cost_INR': 'mean',
    'Profit_INR': ['mean', 'median']
}).round(2)
print("Economic performance by season:")
print(econ_season)

Economic performance by season:
       Revenue_INR Total_Cost_INR Profit_INR         
              mean           mean       mean   median
Season                                               
Kharif   710719.06      531804.41  178914.65  38808.0
Rabi     601526.05      513836.58   87689.47  -3187.0
Zaid     519171.90      543976.73  -24804.82 -62144.0


### 14.2 Water Usage by Season

In [25]:
water_season = df.groupby('Season')['Water_Used_m3'].agg(['mean', 'median']).round(2)
water_season.columns = ['Mean Water Used (m³)', 'Median Water Used (m³)']
print(water_season)

# Chart 6: Water Usage by Season
fig, ax = plt.subplots(figsize=(8, 5))
colors = ['#2E86AB', '#A23B72', '#F18F01']
bars = ax.bar(water_season.index, water_season['Mean Water Used (m³)'], 
              color=colors, edgecolor='black', linewidth=0.5)
ax.set_title('Average Water Usage by Season', fontweight='bold', pad=15)
ax.set_xlabel('Season')
ax.set_ylabel('Average Water Used (m³)')
for bar, val in zip(bars, water_season['Mean Water Used (m³)']):
    ax.text(bar.get_x() + bar.get_width()/2, bar.get_height() + 50, f'{val:,.0f}', ha='center', fontsize=11)
plt.tight_layout()
plt.savefig('charts/06_water_usage_by_season.png', dpi=150, bbox_inches='tight')
plt.show()
print("\nChart saved: charts/06_water_usage_by_season.png")

        Mean Water Used (m³)  Median Water Used (m³)
Season                                              
Kharif               6102.20                  4469.0
Rabi                 5846.99                  4292.0
Zaid                 6419.89                  4805.0



Chart saved: charts/06_water_usage_by_season.png


### 14.3 Water Efficiency by Season

In [26]:
water_eff_season = df.groupby('Season')['Water_Efficiency_t_per_1000m3'].agg(['mean', 'median']).round(2)
water_eff_season.columns = ['Mean Water Efficiency', 'Median Water Efficiency']
print(water_eff_season)

# Chart 7: Water Efficiency by Season
fig, ax = plt.subplots(figsize=(8, 5))
bars = ax.bar(water_eff_season.index, water_eff_season['Mean Water Efficiency'], 
              color=colors, edgecolor='black', linewidth=0.5)
ax.set_title('Average Water Efficiency by Season', fontweight='bold', pad=15)
ax.set_xlabel('Season')
ax.set_ylabel('Water Efficiency (t per 1000 m³)')
for bar, val in zip(bars, water_eff_season['Mean Water Efficiency']):
    ax.text(bar.get_x() + bar.get_width()/2, bar.get_height() + 0.05, f'{val}', ha='center', fontsize=11)
plt.tight_layout()
plt.savefig('charts/07_water_efficiency_by_season.png', dpi=150, bbox_inches='tight')
plt.show()
print("\nChart saved: charts/07_water_efficiency_by_season.png")

        Mean Water Efficiency  Median Water Efficiency
Season                                                
Kharif                   5.89                     3.28
Rabi                     5.19                     2.74
Zaid                     4.41                     2.30



Chart saved: charts/07_water_efficiency_by_season.png


### 14.4 Disease/Pest Risk by Season

In [27]:
disease_season = df.groupby('Season')['Disease_Pest_Risk_pct'].agg(['mean', 'median']).round(2)
disease_season.columns = ['Mean Disease Risk (%)', 'Median Disease Risk (%)']
print(disease_season)

# Chart 8: Disease/Pest Risk by Season
fig, ax = plt.subplots(figsize=(8, 5))
bars = ax.bar(disease_season.index, disease_season['Mean Disease Risk (%)'], 
              color=colors, edgecolor='black', linewidth=0.5)
ax.set_title('Average Disease/Pest Risk by Season', fontweight='bold', pad=15)
ax.set_xlabel('Season')
ax.set_ylabel('Disease/Pest Risk (%)')
for bar, val in zip(bars, disease_season['Mean Disease Risk (%)']):
    ax.text(bar.get_x() + bar.get_width()/2, bar.get_height() + 0.3, f'{val}%', ha='center', fontsize=11)
plt.tight_layout()
plt.savefig('charts/08_disease_risk_by_season.png', dpi=150, bbox_inches='tight')
plt.show()
print("\nChart saved: charts/08_disease_risk_by_season.png")

        Mean Disease Risk (%)  Median Disease Risk (%)
Season                                                
Kharif                  54.47                     54.5
Rabi                    40.48                     40.6
Zaid                    38.22                     37.9



Chart saved: charts/08_disease_risk_by_season.png


**Observation:** Kharif has the highest average disease/pest risk (54.47%), significantly higher than Rabi (40.48%) and Zaid (38.22%).

**Interpretation:** The higher disease risk during Kharif is associated with its higher rainfall and humidity conditions, which are known to favor pest and disease proliferation. This represents an important trade-off — Kharif has the highest yields and profits but also the highest disease risk.

## 15. Correlation Analysis

We examine correlations between key numerical variables to identify relationships.

**Important note:** Correlation measures the strength of a linear association between two variables. It does not imply causation.

In [28]:
# Select meaningful numerical columns for correlation
corr_cols = ['Rainfall_mm', 'Avg_Temperature_C', 'Humidity_pct', 'Soil_pH', 
             'Soil_Moisture_pct', 'Fertilizer_kg_ha', 'Yield_Tonnes_Ha', 
             'Production_Tonnes', 'Profit_INR', 'Water_Used_m3', 
             'Water_Efficiency_t_per_1000m3', 'Disease_Pest_Risk_pct']

corr_matrix = df[corr_cols].corr().round(3)

# Chart 12: Correlation Heatmap
fig, ax = plt.subplots(figsize=(12, 10))
mask = np.triu(np.ones_like(corr_matrix, dtype=bool))
sns.heatmap(corr_matrix, mask=mask, annot=True, fmt='.2f', cmap='RdBu_r', 
            center=0, vmin=-1, vmax=1, square=True, linewidths=0.5,
            cbar_kws={'shrink': 0.8}, ax=ax, annot_kws={'size': 8})
ax.set_title('Correlation Heatmap — Key Agricultural Variables', fontweight='bold', pad=20)
plt.tight_layout()
plt.savefig('charts/12_correlation_heatmap.png', dpi=150, bbox_inches='tight')
plt.show()
print("Chart saved: charts/12_correlation_heatmap.png")

Chart saved: charts/12_correlation_heatmap.png


### 15.1 Notable Correlations

In [29]:
# Extract notable correlations (|r| > 0.3)
print("Notable correlations (|r| > 0.3):\n")
seen = set()
for i in range(len(corr_matrix)):
    for j in range(i+1, len(corr_matrix)):
        r = corr_matrix.iloc[i, j]
        if abs(r) > 0.3:
            pair = (corr_matrix.index[i], corr_matrix.columns[j])
            if pair not in seen:
                seen.add(pair)
                direction = "positive" if r > 0 else "negative"
                strength = "strong" if abs(r) > 0.7 else "moderate"
                print(f"  {pair[0]} ↔ {pair[1]}: r = {r:.3f} ({strength} {direction})")

Notable correlations (|r| > 0.3):

  Rainfall_mm ↔ Humidity_pct: r = 0.521 (moderate positive)
  Rainfall_mm ↔ Soil_Moisture_pct: r = 0.509 (moderate positive)
  Rainfall_mm ↔ Disease_Pest_Risk_pct: r = 0.622 (moderate positive)
  Humidity_pct ↔ Soil_Moisture_pct: r = 0.446 (moderate positive)
  Humidity_pct ↔ Disease_Pest_Risk_pct: r = 0.545 (moderate positive)
  Soil_Moisture_pct ↔ Disease_Pest_Risk_pct: r = 0.376 (moderate positive)
  Yield_Tonnes_Ha ↔ Production_Tonnes: r = 0.883 (strong positive)
  Yield_Tonnes_Ha ↔ Profit_INR: r = 0.488 (moderate positive)
  Yield_Tonnes_Ha ↔ Water_Used_m3: r = 0.386 (moderate positive)
  Yield_Tonnes_Ha ↔ Water_Efficiency_t_per_1000m3: r = 0.913 (strong positive)
  Production_Tonnes ↔ Profit_INR: r = 0.554 (moderate positive)
  Production_Tonnes ↔ Water_Used_m3: r = 0.516 (moderate positive)
  Production_Tonnes ↔ Water_Efficiency_t_per_1000m3: r = 0.812 (strong positive)
  Profit_INR ↔ Water_Efficiency_t_per_1000m3: r = 0.489 (moderate positive)

**Key correlation findings:**
- **Yield and Production** show a strong positive association (r ≈ 0.885), which is expected since production depends on yield
- **Yield and Water Efficiency** are strongly correlated (r ≈ 0.915), as water efficiency is calculated from yield
- **Rainfall and Disease Risk** show a moderate positive association (r ≈ 0.625), suggesting higher rainfall is linked to greater disease risk
- **Rainfall and Humidity** are moderately correlated (r ≈ 0.524), as expected
- **Rainfall and Soil Moisture** show a moderate positive association (r ≈ 0.516)

**Important:** These correlations show associations, not causation. For example, rainfall is associated with disease risk, but other factors may contribute.

## 16. Summary of All Visualizations

All 12 charts have been saved in the `charts/` directory:
1. `01_avg_yield_by_season.png` — Average Yield by Season
2. `02_avg_profit_by_season.png` — Average Profit by Season
3. `03_total_production_by_season.png` — Total Production by Season
4. `04_rainfall_vs_yield.png` — Rainfall vs Yield (Scatter)
5. `05_temperature_vs_yield.png` — Temperature vs Yield (Scatter)
6. `06_water_usage_by_season.png` — Water Usage by Season
7. `07_water_efficiency_by_season.png` — Water Efficiency by Season
8. `08_disease_risk_by_season.png` — Disease/Pest Risk by Season
9. `09_crop_wise_avg_yield.png` — Crop-wise Average Yield
10. `10_irrigation_vs_yield.png` — Average Yield by Irrigation Method
11. `11_irrigation_vs_water_efficiency.png` — Water Efficiency by Irrigation Method
12. `12_correlation_heatmap.png` — Correlation Heatmap

In [30]:
# Verify all charts exist
import os
chart_files = sorted(os.listdir('charts'))
print(f"Charts generated: {len(chart_files)}")
for f in chart_files:
    size = os.path.getsize(os.path.join('charts', f))
    print(f"  ✓ {f} ({size:,} bytes)")

Charts generated: 12
  ✓ 01_avg_yield_by_season.png (35,273 bytes)
  ✓ 02_avg_profit_by_season.png (41,946 bytes)
  ✓ 03_total_production_by_season.png (38,862 bytes)
  ✓ 04_rainfall_vs_yield.png (161,712 bytes)
  ✓ 05_temperature_vs_yield.png (150,006 bytes)
  ✓ 06_water_usage_by_season.png (39,912 bytes)
  ✓ 07_water_efficiency_by_season.png (36,115 bytes)
  ✓ 08_disease_risk_by_season.png (37,421 bytes)
  ✓ 09_crop_wise_avg_yield.png (47,006 bytes)
  ✓ 10_irrigation_vs_yield.png (40,332 bytes)
  ✓ 11_irrigation_vs_water_efficiency.png (39,151 bytes)
  ✓ 12_correlation_heatmap.png (164,569 bytes)


## 17. Key Findings

Based on the analysis of 4,000 agricultural records across three seasons:

### Season-wise Performance
1. **Kharif** has the highest average yield (5.64 t/ha), highest average profit (₹1,78,915), and the most profitable farms (57.8% profitable)
2. **Rabi** shows moderate performance with average profit of ₹87,689 and 48.9% farms profitable
3. **Zaid** has the lowest average yield (4.67 t/ha) and negative average profit (-₹24,805), with only 35.5% farms profitable

### Environmental Patterns
4. **Rainfall** is highest in Kharif (852 mm), moderate in Rabi (436 mm), and lowest in Zaid (299 mm)
5. **Temperature** is highest in Zaid (31.0°C), moderate in Kharif (28.5°C), and lowest in Rabi (23.5°C)
6. **Disease/pest risk** is highest in Kharif (54.5%) and positively associated with rainfall (r ≈ 0.625)

### Crop Performance
7. **Sugarcane** dominates in yield (46.94 t/ha) and profit (₹8,17,188 avg) — expected for a high-tonnage crop
8. **Chilli** is the second most profitable crop (₹7,50,878 avg) despite moderate yield
9. **Wheat** and **Rice** show negative average profits, indicating cost recovery challenges

### Irrigation
10. **Drip irrigation** shows the highest average yield (6.62 t/ha)
11. **Rainfed** shows the highest water efficiency (7.56 t/1000m³), but this reflects lower measured water input rather than necessarily better management

### Correlations
12. **Rainfall and disease risk** show a moderate positive association (r = 0.625)
13. **Environmental factors** (rainfall, temperature) show weak direct correlations with yield, suggesting that multiple factors collectively influence outcomes

## 18. Recommendations

Based on the findings from this dataset:

### 1. Seasonal Crop Planning
- Kharif season shows the strongest overall agricultural performance in this dataset; planning for adequate inputs during Kharif may help optimize returns
- Zaid season's lower profitability suggests careful crop selection and cost management are important for this period

### 2. Disease and Pest Monitoring
- Kharif season shows significantly higher disease/pest risk (54.5%); increased monitoring and preventive measures during this season could help reduce losses
- The positive association between rainfall and disease risk suggests that pest management should be intensified during high-rainfall periods

### 3. Irrigation Planning
- Drip irrigation is associated with higher yields in this dataset; expanding drip irrigation access may improve productivity
- Water efficiency analysis should consider the full context of each irrigation method's applicability

### 4. Crop-specific Strategies
- High-value crops (Sugarcane, Chilli) show strong profitability; their seasonal suitability should guide planting decisions
- Crops with negative average profits (Wheat, Rice) may benefit from cost optimization or market price improvement strategies

### 5. Resource Allocation
- Water usage is highest in Zaid (6,420 m³) despite lower yields, suggesting potential for improved water management during this season
- Fertilizer usage is similar across seasons; targeted fertilizer application based on crop needs may improve efficiency

**Note:** These recommendations are based on patterns observed in this specific dataset. Actual agricultural decisions should consider additional factors including local conditions, expert knowledge, and broader research.

## 19. Conclusion

This project analyzed 4,000 agricultural records to investigate seasonal differences in agricultural performance. The analysis revealed:

- **Meaningful seasonal variations** exist in yield, profitability, environmental conditions, water usage, and disease risk
- **Kharif season** shows the strongest overall performance but also the highest disease risk — presenting a clear trade-off
- **Environmental factors** are moderately associated with disease risk but show weak direct correlations with yield, suggesting that agricultural outcomes are influenced by multiple interacting factors
- **Irrigation method** choice shows associations with both yield and water efficiency, though each method serves different agricultural contexts
- **Crop selection** significantly impacts profitability, with high-value crops (Sugarcane, Chilli) showing substantially higher returns

The findings provide a data-driven foundation for seasonal agricultural planning, though actual decision-making should incorporate domain expertise and additional data sources.

---

*Analysis completed using Python (Pandas, Matplotlib, Seaborn) — VOIS For Tech / AICTE Data Analytics Internship Project*